# Лабораторная работа №3

## Задание 1

In [40]:
import pandas as pd

Таблица рейтингов

In [41]:
ratings_dataframe = pd.read_csv("../lab3/data/ratings.dat", sep="::", names=["UserID", "MovieID", "Rating", "Timestamp"], engine="python")
ratings_dataframe

,UserID,MovieID,Rating,Timestamp
0,1,1193,5,978300760
1,1,661,3,978302109
2,1,914,3,978301968
3,1,3408,4,978300275
4,1,2355,5,978824291
...,...,...,...,...
1000204,6040,1091,1,956716541
1000205,6040,1094,5,956704887
1000206,6040,562,5,956704746
1000207,6040,1096,4,956715648


Таблица пользователей

In [42]:
users_dataframe = pd.read_csv("../lab3/data/users.dat", sep="::", names=["UserID", "Gender", "Age", "Occupation", "Zip-code"], engine="python")
users_dataframe

,UserID,Gender,Age,Occupation,Zip-code
0,1,F,1,10,48067
1,2,M,56,16,70072
2,3,M,25,15,55117
3,4,M,45,7,02460
4,5,M,25,20,55455
...,...,...,...,...,...
6035,6036,F,25,15,32603
6036,6037,F,45,1,76006
6037,6038,F,56,1,14706
6038,6039,F,45,0,01060


Таблица фильмов

In [43]:
movies_dataframe = pd.read_csv("../lab3/data/movies.dat", sep="::", names=["MovieID", "Title", "Genre"], engine="python", encoding="latin-1")
movies_dataframe

,MovieID,Title,Genre
0,1,Toy Story (1995),Animation|Children's|Comedy
1,2,Jumanji (1995),Adventure|Children's|Fantasy
2,3,Grumpier Old Men (1995),Comedy|Romance
3,4,Waiting to Exhale (1995),Comedy|Drama
4,5,Father of the Bride Part II (1995),Comedy
...,...,...,...
3878,3948,Meet the Parents (2000),Comedy
3879,3949,Requiem for a Dream (2000),Drama
3880,3950,Tigerland (2000),Drama
3881,3951,Two Family House (2000),Drama


1) Требуется вычислить средние рейтинги для фильма с ID = № варианта в зависимости от пола;

In [44]:
variant = 48
user_rating_merge_dataframe = pd.merge(users_dataframe, ratings_dataframe, on="UserID")
user_rating_var_dataframe = user_rating_merge_dataframe[user_rating_merge_dataframe["MovieID"] == variant]
female_ratings = user_rating_var_dataframe[user_rating_var_dataframe["Gender"] == "F"]['Rating'].mean()
male_ratings = user_rating_var_dataframe[user_rating_var_dataframe["Gender"] == "M"]['Rating'].mean()
print(f"Средний рейтинг для фильма среди женщин - {round(female_ratings, 2)}\nСредний рейтинг для фильма среди мужчин - {round(male_ratings, 2)}")
movies_dataframe[movies_dataframe["MovieID"] == variant]

Средний рейтинг для фильма среди женщин - 3.16
Средний рейтинг для фильма среди мужчин - 2.85


,MovieID,Title,Genre
47,48,Pocahontas (1995),Animation|Children's|Musical|Romance


2) Отобрать фильмы, получившие не менее 250 оценок, и для них определить средний рейтинг в зависимости от пола;

Результат вывести в виде DataFrame с колонками:

['MovieID', 'Title', 'Rating_F', 'Rating_M']

In [45]:
general_dataframe = pd.merge(user_rating_merge_dataframe, movies_dataframe, on="MovieID")

most_rated = general_dataframe.groupby("MovieID").size()
most_rated = most_rated[most_rated >= 250]

most_rated_dataframe = general_dataframe[general_dataframe["MovieID"].isin(most_rated.index)]
ratings = round(most_rated_dataframe.groupby(["MovieID", "Gender"])['Rating'].mean(), 2).unstack()
ratings_dataframe = pd.merge(ratings, movies_dataframe[["MovieID", "Title"]], on="MovieID")
ratings_dataframe = ratings_dataframe.rename(columns={"F": "Rating_F", "M": "Rating_M"})
ratings_dataframe = ratings_dataframe.reindex(columns=["MovieID", "Title", "Rating_F", "Rating_M"])

display(most_rated_dataframe)
display(ratings_dataframe)
display(general_dataframe)

,UserID,Gender,Age,Occupation,Zip-code,MovieID,Rating,Timestamp,Title,Genre
0,1,F,1,10,48067,1193,5,978300760,One Flew Over the Cuckoo's Nest (1975),Drama
1,1,F,1,10,48067,661,3,978302109,James and the Giant Peach (1996),Animation|Children's|Musical
2,1,F,1,10,48067,914,3,978301968,My Fair Lady (1964),Musical|Romance
3,1,F,1,10,48067,3408,4,978300275,Erin Brockovich (2000),Drama
4,1,F,1,10,48067,2355,5,978824291,"Bug's Life, A (1998)",Animation|Children's|Comedy
...,...,...,...,...,...,...,...,...,...,...
1000204,6040,M,25,6,11106,1091,1,956716541,Weekend at Bernie's (1989),Comedy
1000205,6040,M,25,6,11106,1094,5,956704887,"Crying Game, The (1992)",Drama|Romance|War
1000206,6040,M,25,6,11106,562,5,956704746,Welcome to the Dollhouse (1995),Comedy|Drama
1000207,6040,M,25,6,11106,1096,4,956715648,Sophie's Choice (1982),Drama


,MovieID,Title,Rating_F,Rating_M
0,1,Toy Story (1995),4.19,4.13
1,2,Jumanji (1995),3.28,3.18
2,3,Grumpier Old Men (1995),3.07,2.99
3,5,Father of the Bride Part II (1995),3.21,2.89
4,6,Heat (1995),3.68,3.91
...,...,...,...,...
1211,3917,Hellraiser (1987),3.02,3.13
1212,3927,Fantastic Voyage (1966),3.54,3.59
1213,3948,Meet the Parents (2000),3.62,3.64
1214,3949,Requiem for a Dream (2000),3.95,4.17


,UserID,Gender,Age,Occupation,Zip-code,MovieID,Rating,Timestamp,Title,Genre
0,1,F,1,10,48067,1193,5,978300760,One Flew Over the Cuckoo's Nest (1975),Drama
1,1,F,1,10,48067,661,3,978302109,James and the Giant Peach (1996),Animation|Children's|Musical
2,1,F,1,10,48067,914,3,978301968,My Fair Lady (1964),Musical|Romance
3,1,F,1,10,48067,3408,4,978300275,Erin Brockovich (2000),Drama
4,1,F,1,10,48067,2355,5,978824291,"Bug's Life, A (1998)",Animation|Children's|Comedy
...,...,...,...,...,...,...,...,...,...,...
1000204,6040,M,25,6,11106,1091,1,956716541,Weekend at Bernie's (1989),Comedy
1000205,6040,M,25,6,11106,1094,5,956704887,"Crying Game, The (1992)",Drama|Romance|War
1000206,6040,M,25,6,11106,562,5,956704746,Welcome to the Dollhouse (1995),Comedy|Drama
1000207,6040,M,25,6,11106,1096,4,956715648,Sophie's Choice (1982),Drama


3) Найти фильмы, оказавшиеся в топ 5 у зрителей-женщин и худшие 5 у зрителей- мужчин. Найти фильмы (топ 5), по которым мужчины и женщины сильнее всего разошлись в оценках;

In [46]:
top_five_F = ratings_dataframe.sort_values(by="Rating_F", ascending=False)[:5]
display(top_five_F)

worst_five_M = ratings_dataframe.sort_values(by="Rating_M", ascending=True)[:5]
display(worst_five_M)

ratings_dataframe["Difference"] = ratings_dataframe.apply(lambda x: abs(x["Rating_F"] - x["Rating_M"]), axis=1)
top_five_dif = ratings_dataframe.sort_values(by="Difference", ascending=False)[:5]
display(top_five_dif)

,MovieID,Title,Rating_F,Rating_M
212,745,"Close Shave, A (1995)",4.64,4.47
336,1148,"Wrong Trousers, The (1993)",4.59,4.48
255,922,Sunset Blvd. (a.k.a. Sunset Boulevard) (1950),4.57,4.46
206,720,Wallace & Gromit: The Best of Aardman Animatio...,4.56,4.39
162,527,Schindler's List (1993),4.56,4.49


,MovieID,Title,Rating_F,Rating_M
1139,3593,Battlefield Earth (2000),1.57,1.62
465,1381,Grease 2 (1982),2.24,1.79
172,546,Super Mario Bros. (1993),2.16,1.82
841,2643,Superman IV: The Quest for Peace (1987),2.22,1.85
502,1556,Speed 2: Cruise Control (1997),1.91,1.86


,MovieID,Title,Rating_F,Rating_M,Difference
313,1088,Dirty Dancing (1987),3.79,2.96,0.83
352,1201,"Good, The Bad and The Ugly, The (1966)",3.49,4.22,0.73
1186,3760,"Kentucky Fried Movie, The (1977)",2.88,3.56,0.68
796,2468,Jumpin' Jack Flash (1986),3.25,2.58,0.67
68,231,Dumb & Dumber (1994),2.70,3.34,0.64


4) Определить фильмы (топ 5), вызвавшие в среднем наибольшее разногласие у зрителей независимо от пола;

In [47]:
rating_std = general_dataframe.groupby("MovieID")["Rating"].std().round(2)
rating_std.name = "STD"
rating_std_dataframe = pd.merge(rating_std, movies_dataframe[["MovieID", "Title"]], on="MovieID")
rating_std_dataframe = rating_std_dataframe.reindex(columns=["MovieID", "Title", "STD"])

rating_std_dataframe = rating_std_dataframe.sort_values("STD", ascending=False).head(5)
display(rating_std_dataframe)

,MovieID,Title,STD
558,572,Foreign Student (1994),2.83
3557,3800,Criminal Lovers (Les Amants Criminels) (1999),2.31
1262,1360,Identification of a Woman (Identificazione di ...,2.12
676,706,Sunset Park (1996),2.12
1612,1773,Tokyo Fist (1995),2.12


5) Определить средний рейтинг каждого жанра для различных возрастов. Установите, фильмы какого жанра в каждой категории возрастов имеют наибольший и наименьший рейтинги. Если фильм относится к нескольким данным, то их надо разделить и сформировать несколько строчек с этим фильмом, но с различными жанрами, например:

| UserID<br> | MovieID | Rating | Timestamp | Gender | Age | Occupation | Zip-code |              Title               |               Genres               |
| :--------: | :-----: | :----: | :-------: | :----: | :-: | :--------: | :------: | :------------------------------: | :--------------------------------: |
|     1      |   661   |   3    | 978302109 |   F    |  1  |     10     |  48067   | James and the Giant Peach (1996) | Animation \| Children's \| Musical |

необходимо преобразовать в

| UserID | MovieID | Rating | Timestamp | Gender | Age | Occupation | Zip-code |              Title               |   Genres   |
| :----: | :-----: | :----: | :-------: | :----: | :-: | :--------: | :------: | :------------------------------: | :--------: |
|   1    |   661   |   3    | 978302109 |   F    |  1  |     10     |  48067   | James and the Giant Peach (1996) | Animation  |
|   1    |   661   |   3    | 978302109 |   F    |  1  |     10     |  48067   | James and the Giant Peach (1996) | Children's |
|   1    |   661   |   3    | 978302109 |   F    |  1  |     10     |  48067   | James and the Giant Peach (1996) |  Musical   |


In [48]:
general_dataframe["Genre"] = general_dataframe["Genre"].str.split("|")
exploded_general_dataframe = general_dataframe.explode("Genre")
genre_ratings = round(exploded_general_dataframe.groupby(["Genre", "Age"])["Rating"].mean(), 2).reset_index()
print("Средняя оценка жанров по возрасту")
display(genre_ratings)

max_rating_by_age = genre_ratings.loc[genre_ratings.groupby("Age")["Rating"].idxmax()]
min_rating_by_age = genre_ratings.loc[genre_ratings.groupby("Age")["Rating"].idxmin()]

print("Максимальные оценки по возрастам:")
display(max_rating_by_age)

print("Минимальные оценки по возрастам:")
display(min_rating_by_age)

Средняя оценка жанров по возрасту


,Genre,Age,Rating
0,Action,1,3.51
1,Action,18,3.45
2,Action,25,3.45
3,Action,35,3.54
4,Action,45,3.53
...,...,...,...
121,Western,25,3.61
122,Western,35,3.68
123,Western,45,3.67
124,Western,50,3.74


Максимальные оценки по возрастам:


,Genre,Age,Rating
63,Film-Noir,1,4.15
64,Film-Noir,18,4.00
65,Film-Noir,25,4.06
66,Film-Noir,35,4.06
67,Film-Noir,45,4.11
68,Film-Noir,50,4.18
69,Film-Noir,56,4.13


Минимальные оценки по возрастам:


,Genre,Age,Rating
21,Children's,1,3.24
71,Horror,18,3.17
72,Horror,25,3.20
73,Horror,35,3.28
74,Horror,45,3.26
75,Horror,50,3.16
76,Horror,56,3.25


### задание 1 для защиты
вывести фильмы, которые мужчины смотрели больше в 2 раза, чем женщины

In [ ]:
counts = general_dataframe.groupby(['MovieID', 'Gender']).size()
print(counts)
counts_unstacked = counts.unstack()
print(counts_unstacked)

male_double_females_df = counts_unstacked[counts_unstacked['M'] >= 2 * counts_unstacked['F']]
display(male_double_females_df)

result_final_df = pd.merge(male_double_females_df, movies_dataframe[['MovieID', 'Title']], on='MovieID')

result_final_df = result_final_df.sort_values(by='MovieID')
display(result_final_df)

MovieID  Gender
1        F          591
         M         1486
2        F          176
         M          525
3        F          136
                   ... 
3950     M           44
3951     F           17
         M           23
3952     F          105
         M          283
Length: 7152, dtype: int64
Gender       F       M
MovieID               
1        591.0  1486.0
2        176.0   525.0
3        136.0   342.0
4         85.0    85.0
5        108.0   188.0
...        ...     ...
3948     231.0   631.0
3949      80.0   224.0
3950      10.0    44.0
3951      17.0    23.0
3952     105.0   283.0

[3706 rows x 2 columns]


Gender,F,M
MovieID,,
1,591.0,1486.0
2,176.0,525.0
3,136.0,342.0
6,129.0,811.0
9,10.0,92.0
...,...,...
3947,5.0,50.0
3948,231.0,631.0
3949,80.0,224.0


,MovieID,Title,F,M
0,1,Toy Story (1995),591.0,1486.0
1,2,Jumanji (1995),176.0,525.0
2,3,Grumpier Old Men (1995),136.0,342.0
3,6,Heat (1995),129.0,811.0
4,9,Sudden Death (1995),10.0,92.0
...,...,...,...,...
2579,3947,Get Carter (1971),5.0,50.0
2580,3948,Meet the Parents (2000),231.0,631.0
2581,3949,Requiem for a Dream (2000),80.0,224.0
2582,3950,Tigerland (2000),10.0,44.0
